In [ ]:
# Load necessary libraries
import sys
import os

sys.path.append(os.path.abspath(".."))

import pandas as pd 
from sklearn.model_selection import train_test_split

from src.preprocessing import build_preprocessor
from src.models import build_decision_tree
from src.evaluation import evaluate_model

In [ ]:
# Load the data
df = pd.read_csv("data/WA_Fn-UseC_-Telco-Customer-Churn.csv")

# Preprocess the data
df['TotalCharges'] = pd.to_numeric(df['TotalCharges'], errors='coerce')
df = df.dropna()

df['Churn'] = df['Churn'].map({'Yes': 1, 'No': 0})
df = df.drop(columns=['customerID'])

In [ ]:
# Split the data
X = df.drop(columns='Churn')
y = df['Churn']

X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y
)

In [ ]:
# Build the model
preprocessor = build_preprocessor(X)
model = build_decision_tree(preprocessor)

model.fit(X_train, y_train)

In [ ]:
# Evaluate the model
y_pred = model.predict(X_test)
y_proba = model.predict_proba(X_test)[:, 1]

evaluate_model(y_test, y_pred, y_proba)

Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.85      0.84      1033
           1       0.57      0.55      0.56       374

    accuracy                           0.77      1407
   macro avg       0.70      0.70      0.70      1407
weighted avg       0.77      0.77      0.77      1407

ROC AUC Score: 0.7933048956623925
Confusion Matrix:
 [[879 154]
 [169 205]]


First run with depth 5:

Classification Report:
              precision    recall  f1-score   support

           0       0.86      0.85      0.86      1033
           1       0.60      0.61      0.61       374

    accuracy                           0.79      1407
   macro avg       0.73      0.73      0.73      1407
weighted avg       0.79      0.79      0.79      1407

ROC AUC Score: 0.8296250472379394
Confusion Matrix:
 [[881 152]
 [144 230]]


Second run depth 6:

Classification Report:
              precision    recall  f1-score   support

           0       0.85      0.87      0.86      1033
           1       0.62      0.56      0.59       374

    accuracy                           0.79      1407
   macro avg       0.73      0.72      0.72      1407
weighted avg       0.78      0.79      0.79      1407

ROC AUC Score: 0.8231709211009934
Confusion Matrix:
 [[903 130]
 [165 209]]

Third run depth 8:

Classification Report:
              precision    recall  f1-score   support

           0       0.84      0.85      0.84      1033
           1       0.57      0.55      0.56       374

    accuracy                           0.77      1407
   macro avg       0.70      0.70      0.70      1407
weighted avg       0.77      0.77      0.77      1407

ROC AUC Score: 0.7933048956623925
Confusion Matrix:
 [[879 154]
 [169 205]]

First conclusion: The Decision Tree improves recall, but slightly worsens overall ranking quality (ROC-AUC).

Increasing tree depth leads to overfitting, where the model captures noise rather than general patterns. A depth of 5 provides the best balance between bias and variance, improving recall for churners while maintaining reasonable generalization.

The Decision Tree model improves recall for churn prediction compared to Logistic Regression, identifying more at-risk customers. However, increasing tree depth leads to overfitting and reduced performance on unseen data. A depth of 5 provides the best trade-off between model complexity and generalization.